src/io/category_encoding_provider.cpp

```diff
@@ -0,0 +1,1330 @@
+/*!
+  * Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+  * Licensed under the MIT License. See LICENSE file in the project root for license information.
+  */
+
+#include <LightGBM/category_encoding_provider.hpp>
+
+#include <queue>
+#include <set>
+
+namespace LightGBM {
+
+CategoryEncodingProvider::CategoryEncodingProvider(Config* config) {
```

Comment:
Should we have a unified construction for CategoryEncodingProvider? And others are just transform inputs into this construction method? like below:

CategoryEncodingProvider(a, b, c, d, e)

CategoryEncodingProvider(a, b, c, d) {
CategoryEncodingProvider((a, b, c, d, 0)
}